### PR TITLE
tzc400: allow platform to check TZC400 configuration and use in plat-stm32mp1

### PR DESCRIPTION
**Pull request by etienne-lms**

### CODE REVIEW

File: core/drivers/tzc400.c [hunk 1]
```diff
@@ -284,6 +314,27 @@ void tzc_configure_region(uint32_t filters,
...
+
+	cfg->base = reg_pair_to_64(tzc_read_region_base_high(tzc.base, region),
+				   tzc_read_region_base_low(tzc.base, region));
+	cfg->top = reg_pair_to_64(tzc_read_region_top_high(tzc.base, region),
+				  tzc_read_region_top_low(tzc.base, region));
+
+	cfg->ns_device_access = tzc_read_region_id_access(tzc.base, region);
+
+	val32 = tzc_read_region_attributes(tzc.base, region);
+	cfg->sec_attr = (enum tzc_region_attributes)(val32 >>
```
Comment (jenswi-linaro) [inline core/drivers/tzc400.c:324]: Cast is not needed.

File: core/drivers/tzc400.c [hunk 2]
```diff
@@ -391,6 +391,73 @@ void tzc_disable_filters(void)
...
+	EMSG("Violation @0x%"PRIx64", %ssecure %sprivileged %s, AXI ID %"PRIx32,
+	     addr,
+	     secure_not_nonsecure(filter) ? "" : "non-",
+	     priv_not_unpriv(filter) ? "" : "un",
+	     read_not_write(filter) ? "read" : "write",
+	     io_read32(tzc.base + FAIL_ID(filter)));
+}
+
+/*
+ * Dump info when TZC400 catchs an unallowed access with TZC
```
Comment (jenswi-linaro) [inline core/drivers/tzc400.c:435]: s/catchs/catches/

File: core/include/drivers/tzc400.h [hunk 3]
```diff
@@ -138,15 +145,15 @@
...
-#define FAIL_CONTROL_DIR_SHIFT		(1 << 24)
-#define  FAIL_CONTROL_DIR_READ		0x0
-#define  FAIL_CONTROL_DIR_WRITE		0x1
-#define FAIL_CONTROL_NS_SHIFT		(1 << 21)
-#define  FAIL_CONTROL_NS_SECURE		0x0
-#define  FAIL_CONTROL_NS_NONSECURE	0x1
-#define FAIL_CONTROL_PRIV_SHIFT		(1 << 20)
-#define  FAIL_CONTROL_PRIV_PRIV		0x0
-#define  FAIL_CONTROL_PRIV_UNPRIV	0x1
+#define FAIL_CONTROL_DIR		BIT(24)
```
Comment (jenswi-linaro) [inline core/include/drivers/tzc400.h:148]: How about defining it as:
``` C
/* If set write access, else read access */
#define FAIL_CONTROL_DIRECTION_WRITE BIT(24)
/* If set non-secure access, else secure access */
#define FAIL_CONTROL_NONSECURE       BIT(21)
/* If set privileged access, else unprivileged access */
#define FAIL_CONTROL_PRIVILEGED      BIT(20)
```

Easier to use since it only involves one define for checking each bit.

File: core/drivers/tzc400.c [hunk 4]
```diff
@@ -262,56 +262,48 @@ static uint32_t addr_high(vaddr_t addr __unused)
  * this cannot be changed. It is, however, possible to change some region 0
  * permissions.
  */
-void tzc_configure_region(uint32_t filters,
-			  uint8_t  region,
-			  vaddr_t  region_base,
-			  vaddr_t  region_top,
-			  enum tzc_region_attributes sec_attr,
-			  uint32_t ns_device_access)
+void tzc_configure_region(uint8_t region, struct tzc_region_config *cfg)
```
Comment (jenswi-linaro) [inline core/drivers/tzc400.c:265]: `const struct tzc_region_config`

File: core/arch/arm/plat-stm32mp1/plat_tzc400.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	if (tzc_get_region_config(i, &region_cfg))
+		panic();
+
+	return region_cfg.base == base && region_cfg.top == (base + size - 1) &&
+	       region_cfg.sec_attr == TZC_REGION_S_RDWR &&
+	       region_cfg.ns_device_access == 0 &&
+	       region_cfg.filters == filters_mask;
+}
+
+#define _DRAM_BEGIN		DDR_BASE
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm32mp1/plat_tzc400.c:65]: Please make these local variables inside `init_stm32mp1_tzc()` instead.

File: core/arch/arm/plat-stm32mp1/main.c [hunk 6]
```diff
@@ -54,17 +54,18 @@ register_phys_mem_pgdir(MEM_AREA_IO_SEC, PWR_BASE, SMALL_PAGE_SIZE);
...
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, TZC_BASE, SMALL_PAGE_SIZE);
 register_phys_mem_pgdir(MEM_AREA_IO_SEC, USART1_BASE, SMALL_PAGE_SIZE);
 
 #if DDR_BASE < CFG_TZDRAM_START
 register_dynamic_shm(DDR_BASE, CFG_TZDRAM_START - DDR_BASE);
 #endif
 
-#define DRAM_END		(DDR_BASE + CFG_DRAM_SIZE)
-#define TZDRAM_END		(CFG_TZDRAM_START + CFG_TZDRAM_SIZE)
+#define _DRAM_END		(DDR_BASE + CFG_DRAM_SIZE)
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm32mp1/main.c:64]: This doesn't seem to be related to the commit description.

File: core/drivers/tzc400.c (hunk 1 above)
Comment (etienne-lms) [reply core/drivers/tzc400.c:324]: ok

File: core/drivers/tzc400.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/tzc400.c:435]: thanks

File: core/include/drivers/tzc400.h (hunk 3 above)
Comment (etienne-lms) [reply core/include/drivers/tzc400.h:148]: much simpler. I will update. thanks.

File: core/drivers/tzc400.c (hunk 4 above)
Comment (etienne-lms) [reply core/drivers/tzc400.c:265]: ok

File: core/arch/arm/plat-stm32mp1/main.c (hunk 6 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/main.c:64]: True. This is related my previous implementation where `init_stm32mp1_tzc()` was implemented in **main.c**. 
I will discard this change.

File: core/arch/arm/plat-stm32mp1/plat_tzc400.c (hunk 5 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/plat_tzc400.c:65]: ok

**Comment by etienne-lms** [comment]
Rebased.

For commit "tzc400: add API functions to get firewall configuration":
Cast removed.
Tag applied.

For commit "tzc400: remove useless traces from tzc_dump_state()":
Tag applied.

For commit "tzc400: new api functions to print violation on device interrupt":
Comments addressed in a fixup commit.

For commit "tzc400: simply tzc_configure_region() arguments"
Comments addressed.
Tag applied.

Comments for commit "plat-stm32mp1: check TZC400 configuration"
Comments addressed in a fixup commit.

File: core/arch/arm/plat-stm32mp1/plat_tzc400.c [hunk 7]
```diff
@@ -84,6 +77,13 @@ static TEE_Result init_stm32mp1_tzc(void)
 	 * OP-TEE hence here only check this complies with static Core
 	 * expectations.
 	 */
+#define _DRAM_BEGIN		DDR_BASE
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm32mp1/plat_tzc400.c:80]: Why are you not using local variables instead?

**Comment by jenswi-linaro** [comment]
For commit "tzc400: new api functions to print violation on device interrupt" with fixup:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/plat-stm32mp1/plat_tzc400.c (hunk 7 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/plat_tzc400.c:80]: fair :)

**Comment by etienne-lms** [comment]
updated

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
fixup commits squashed.
review tags applied on all commits.
